pkg/sessions/tests/session_store_tests.go

```diff
@@ -0,0 +1,435 @@
+				cookie := cookiesapi.MakeCookieFromOptions(in.request, in.cookieOpts.Name, value, in.cookieOpts, in.cookieOpts.Expire, time.Now())
+				in.request.AddCookie(cookie)
+
+				err := in.ss().Save(in.response, in.request, in.session)
+				Expect(err).ToNot(HaveOccurred())
+			})
+
+			It("sets a `set-cookie` header in the response", func() {
+				Expect(in.response.Header().Get("set-cookie")).ToNot(BeEmpty())
+			})
+
+			It("Ensures the session CreatedAt is not zero", func() {
+				Expect(in.session.CreatedAt.IsZero()).To(BeFalse())
+			})
+
+			CheckCookieOptions(in)
+		})
+
+		Context("with an expired saved session", func() {
+			var err error
+			BeforeEach(func() {
+				By("saving a session")
+				req := httptest.NewRequest("GET", "http://example.com/", nil)
+				saveResp := httptest.NewRecorder()
+				err = in.ss().Save(saveResp, req, in.session)
+				Expect(err).ToNot(HaveOccurred())
+
+				By("and clearing the session")
+				for _, c := range saveResp.Result().Cookies() {
+					in.request.AddCookie(c)
+				}
+				clearResp := httptest.NewRecorder()
+				err = in.ss().Clear(clearResp, in.request)
+				Expect(err).ToNot(HaveOccurred())
+
+				By("then saving a request with the cleared session")
+				err = in.ss().Save(in.response, in.request, in.session)
+			})
+
+			It("no error should occur", func() {
+				Expect(err).ToNot(HaveOccurred())
+			})
+		})
+	})
+
+	Context("when Clear is called", func() {
+		BeforeEach(func() {
+			req := httptest.NewRequest("GET", "http://example.com/", nil)
+			saveResp := httptest.NewRecorder()
+			err := in.ss().Save(saveResp, req, in.session)
+			Expect(err).ToNot(HaveOccurred())
+
+			for _, c := range saveResp.Result().Cookies() {
+				in.request.AddCookie(c)
+			}
+			err = in.ss().Clear(in.response, in.request)
+			Expect(err).ToNot(HaveOccurred())
+		})
+
+		It("sets a `set-cookie` header in the response", func() {
+			Expect(in.response.Header().Get("Set-Cookie")).ToNot(BeEmpty())
+		})
+
+		CheckCookieOptions(in)
+	})
+
+	Context("when Load is called", func() {
+		BeforeEach(func() {
+			req := httptest.NewRequest("GET", "http://example.com/", nil)
+			resp := httptest.NewRecorder()
+			err := in.ss().Save(resp, req, in.session)
+			Expect(err).ToNot(HaveOccurred())
+
+			for _, cookie := range resp.Result().Cookies() {
+				in.request.AddCookie(cookie)
+			}
+		})
+
+		Context("before the refresh period", func() {
+			LoadSessionTests(in)
+		})
+
+	})
+}
+
+func LoadSessionTests(in *testInput) {
+	var loadedSession *sessionsapi.SessionState
+	BeforeEach(func() {
+		var err error
+		loadedSession, err = in.ss().Load(in.request)
+		Expect(err).ToNot(HaveOccurred())
+	})
+
+	It("loads a session equal to the original session", func() {
+		// Can't compare time.Time using Equal() so remove ExpiresOn from sessions
+		l := *loadedSession
+		l.CreatedAt = nil
+		l.ExpiresOn = nil
+		s := *in.session
+		s.CreatedAt = nil
+		s.ExpiresOn = nil
+		Expect(l).To(Equal(s))
+
+		// Compare time.Time separately
```

Comment:
This is the last area I had a question about:

We seem to be setting this to `nil` above and worthlessly testing `nil == nil`.  I think we should test times above (via `.Unix()` to smooth out any small format changes) before setting them to `nil` to allow the smooth `Expect(l).To(Equal(s))`

Reply:
Because i'm taking a dereference above `l := *loadedSession`, I'm removing the timestamp from a copy of the loaded sessions not the actual sessions which is where the time is being compared from. I can move it around or check they're non nil if you think that would be more obvious? (I have tested both way arounds and they pass both ways)

As for using `.Unix()`, I don't think we need to since time's have their own [comparator](https://golang.org/pkg/time/#Time.Equal) built in that we are using here, it should be sufficient I think

Reply:
I'll need to revisit mine and see if I can get time comparisons in tests to work without `.Unix()` -- without it here I was having issues (they were ever so slightly different):

https://github.com/oauth2-proxy/oauth2-proxy/pull/632/files#diff-6be19ba25b1420fa1bf80c48dd127f84R256

Reply:
> Because i'm taking a dereference above `l := *loadedSession`, I'm removing the timestamp from a copy of the loaded sessions not the actual sessions which is where the time is being compared from. I can move it around or check they're non nil if you think that would be more obvious? (I have tested both way arounds and they pass both ways)
> 
> As for using `.Unix()`, I don't think we need to since time's have their own [comparator](https://golang.org/pkg/time/#Time.Equal) built in that we are using here, it should be sufficient I think

This is what happens when you go 10 years without working in a language without pointers & memory dereferencing.  Your memory gets fuzzy.  Thanks python 😦

Reply:
> I'll need to revisit mine and see if I can get time comparisons in tests to work without `.Unix()` -- without it here I was having issues (they were ever so slightly different):
> 
> https://github.com/oauth2-proxy/oauth2-proxy/pull/632/files#diff-6be19ba25b1420fa1bf80c48dd127f84R256

Give it a go! I haven't seen this fail when running these tests, on my machine or in CI, so it seems to be pretty stable, can try and make it fail if you like? Alternatively, if we merge and then start seeing issues I'll just turn it off or convert to Unix when we see that happen

Reply:
Weird -- no luck for me without `.Unix()` --

```
Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -1,2 +1,2 @@
            	            	-(*time.Time)(2020-07-01 14:47:00.794741 -0700 PDT)
            	            	+(*time.Time)(2020-07-01 14:47:00.794741 -0700 PDT m=+0.001583017)
```

Reply:
You'll need to do `assert.Equal(t, true, time1.Equals(time2))` rather than just 1 time equals another (which is what that assert error looks like to me)

Reply:
(Not this test - on my own branch. This test passes for me)